[INTLY-5214] Check all alerts as part of e2e & functional tests

dimitraz opened the pull request:
### Description 
Check all alerts exist as part of e2e and functional tests 
- Alerts we expected that weren't found 
- Alerts that were found and not expected 

### Verification
- Log in to an existing Openshift 4 cluster 
- Install integreatly from master
- Check out this branch and run `make test/functional`. The test is common to both e2e and functional tests, so running the functional ones is just quicker 
- See that the output of the test is as expected (e.g. try comment out some of the expected alerts, verify that the report output is correct)

aidenkeating commented on test/common/alerts_exist.go:
```diff
@@ -0,0 +1,281 @@
+	var promApiCallOutput apiResponse
+	err = json.Unmarshal([]byte(output), &promApiCallOutput)
+	if err != nil {
+		t.Logf("Failed to unmarshall json: %s", err)
```
Will this allow the test to go past L183 if it cannot unmarshal the response, is this what we want?

aidenkeating commented on test/common/alerts_exist.go:
```diff
@@ -0,0 +1,281 @@
+	var rulesResult prometheusv1.RulesResult
+	err = json.Unmarshal([]byte(promApiCallOutput.Data), &rulesResult)
+	if err != nil {
+		t.Logf("Failed to unmarshall json: %s", err)
```
Same question as above here

aidenkeating commented on test/common/alerts_exist.go:
```diff
@@ -0,0 +1,281 @@
+		for _, group := range rulesResult.Groups {
+			if !strings.Contains(group.File, file) {
+				continue
+			}
```
Is it expected that there'd only be one `file`-`group.File` match per group or could we get to L200 more than once per group here?

Also, I'm guessing we're using `strings#Contains` here because `group.File` isn't exactly the same as `file`?

aidenkeating commented on test/common/types.go:
```diff
@@ -19,3 +27,11 @@ type TestCase struct {
 	Test        func(t *testing.T, ctx *TestingContext)
 }
+
+type apiResponse struct {
```
If this is going to be stored in the `common` package we should probably make the name of this clearer so it's clear what it's an API response from. wdyt?

dimitraz replied:
Yes good point!

aidenkeating commented on test/common/alerts_exist.go:
```diff
@@ -0,0 +1,281 @@
+)
+
+var expectedRules = map[string][]string{
+	"/redhat-rhmi-middleware-monitoring": {
```
Can we sync up on the structure of these rules? I'd like to understand this structure some more, mainly out of curiosity.

dimitraz replied:
Sure

aidenkeating commented on test/common/alerts_exist.go:
```diff
@@ -0,0 +1,281 @@
+
+	// exec into the prometheus pod
+	output, err := execToPod("curl localhost:9090/api/v1/rules",
+		"prometheus-application-monitoring-0",
```
Are we confident this Pod name should/will always be around?

dimitraz commented on test/common/alerts_exist.go:
```diff
@@ -0,0 +1,281 @@
+		//"ComponentHealth",
+		//"PendingPods",
+		//"RestartingPods",
+		//"TerminatingPods",
```
Yep, will just need to remember to add them back in when the amq online version upgrade has been merged

dimitraz commented:
/retest

dimitraz commented:
Hey @wei-lee, having troubles with the order of the e2e tests being run. `common` tests are currently executed first, but the test I've added relies on the cluster being set up properly. For now I've changed the order of execution but would be nice for the `crd exists` common test to fail sooner rather than at the end. 
What would your recommendation here be? 

cc @david-martin

wei-lee commented:
@dimitraz sorry for the delay, just saw your message. Yes, absolutely, we should change the order for now. I think we are doing too much in `IntegreatlyCluster` for now, ideally many of the tests should be moved out into the `common` directory so that they should become test cases, but it is outside the scope of your PR.

The `crd exists` is really just an example, we don't really need to keep it if it's not required. If we want it to fail sooner, one thing we can look at is to have different stages for the tests and run them at different points. E.g. we could have tests that can run before install, after operator install, after products install etc.  If we do this way, we can split `ALL_TESTS` into a few arrays and invoke them at different point.

Wdyt?

b1zzu commented on test/common/alerts_exist.go:
```diff
@@ -0,0 +1,530 @@
+	}
+
+	// exec into the prometheus pod
+	output, err := execToPod("curl localhost:9090/api/v1/rules",
```
Wouldn't it be better to use `port-forward` to create a local proxy to the pod and then hit it using the golang `http` instant of executing curl inside the pod?

dimitraz replied:
I can look into it, why would this be a better way of doing it rather than doing an exec?

david-martin replied:
I would question the benefit in changing this implementation without any benefit. Using port-forward sounds more cumbersome and prone to error also

b1zzu replied:
I don't have strong opinion for this, but I think that if we could reach the api directly it would simplify the code, as we would not have to deal with streams, stdout and stderr and the `execToPod` method wouldn't be necessary anymore but there is no easy way to use `port-forwarding` too

b1zzu replied:
Would like to mark this ad resolved but somehow I can't, same for the other two comments

b1zzu commented on test/common/alerts_exist.go:
```diff
@@ -0,0 +1,530 @@
+	err := ctx.Client.Get(goctx.TODO(), types.NamespacedName{Name: installationName, Namespace: ns}, rhmi)
+	if err != nil {
+		t.Fatal("error getting RHMI CR:", err)
+	}
```
This information should be a test parameter or flag (don't know what can be used in golang), and this would be a problem of many other tests so it should be solved globally

ping @wei-lee

wei-lee replied:
@b1zzu can be elaborate on what exactly do you mean here?

b1zzu replied:
The code here is checking if the cluster is using or not AWS resources, but when we run the test pipeline we already know if the cluster we have provisioned should use or not the aws resources, therefore we should pass this information to the tests instant of the tests finding it out

david-martin replied:
I think checking the CR is valid. It's the system of record for what type of storage is being used.
No need to double up on this. The test would instruct what to set to the CR to, not the other way around.

b1zzu commented on test/common/alerts_exist.go:
```diff
@@ -0,0 +1,530 @@
+	}
+
+	if missingCount > 0 {
+		fmt.Println("\nMissing alerts were found from Prometheus. If the removal of these Alert rules was intentional, please update this test to remove them from the check. If the removal of these Alert rules was not intendended or you are not sure, please create a Jira & discuss with the monitoring team on how best to proceed")
```
It would be good to print which alerts are missing or are extra

dimitraz replied:
Yep that happens here: https://github.com/integr8ly/integreatly-operator/pull/438/files#diff-fb2239bb04ca074b4768a268fef37e9aR421-R424

b1zzu replied:
Ok thanks, didn't notice it

dimitraz commented:
/retest

dimitraz commented:
/test vendor

dimitraz commented:
/test manifest

dimitraz commented:
/retest

david-martin commented:
/approve

grdryn commented:
/lgtm
/approve

philbrookes commented:
/hold currently prioritising work that is blocking the release

aidenkeating commented on test/common/alerts_exist.go:
```diff
@@ -0,0 +1,536 @@
+}
+
+// difference returns the elements in list of alertsTestRule `a` that aren't in list `b`.
+func ruleDifference(a, b []alertsTestRule) []alertsTestRule {
```
Not important as the comment covers what it does. But I think the var names here could be more descriptive.

For example `mb`, I think means `map representation of  b`.

```
//ruleDifference one-way diff that return rules in diffSource that are not in diffTarget
ruleDifference(diffTarget, diffSource []alertsTestRule) []alertsTestRule {
  //lookup map to check diffSource contains element simple
  diffSourceLookupMap := make(map[string]struct{}, len(diffSource))
  for _, rule := range diffSource {
    diffSourceLookupMap[rule.File] = struct{}{}
  }
  //store files in diffSource that are not in diffTarget
  var diff []alertsTestRule
  for _, rule := range diffTarget {
    if _, found := diffSourceLookupMap[rule.File]; !found {
      diff = append(diff, rule)
    }
  }
  return diff
}
```

I'd then also be tempted to swap source and target.

Again, not very important, just thought I'd mention it.

dimitraz replied:
Thanks Aiden, I agree, that is **a lot** more readable! Will update now

jhellar commented:
Haven't yet had a chance to verify, but code LGTM. Just two things that maybe are not part of this piece of work: tests are currently not executed with `make test/functional`, also there is no test report in console.

dimitraz commented:
fixed, thanks @jhellar

matskiv commented:
/unhold

david-martin commented:
/lgtm
